src/vs/workbench/api/node/extHostTerminalService.ts

```diff
@@ -377,24 +409,37 @@ export class ExtHostTerminalService implements ExtHostTerminalServiceShape {
 
 	public $acceptTerminalClosed(id: number): void {
 		const index = this._getTerminalObjectIndexById(this.terminals, id);
-		if (index === null) {
-			return;
+		if (index !== null) {
+			const terminal = this._terminals.splice(index, 1)[0];
+			this._onDidCloseTerminal.fire(terminal);
 		}
-		const terminal = this._terminals.splice(index, 1)[0];
-		this._onDidCloseTerminal.fire(terminal);
 	}
 
-	public $acceptTerminalOpened(id: number, name: string): void {
-		const index = this._getTerminalObjectIndexById(this._terminals, id);
+	public $acceptTerminalOpened(id: number, name: string, isRendererOnly: boolean, cols: number, rows: number): void {
+		let index = this._getTerminalObjectIndexById(this._terminals, id);
 		if (index !== null) {
-			// The terminal has already been created (via createTerminal*), only fire the event
 			this._onDidOpenTerminal.fire(this.terminals[index]);
+		}
+
+		let renderer = this._getTerminalRendererById(id);
+
+		// If this is a terminal created by one of the public createTerminal* APIs
+		// then @acceptTerminalOpened was called from the main thread task
+		// to indicate that the terminal is ready for use.
+		// In those cases, we don't need to create the extension host objects
+		// as they already exist, but, we do still need to fire the events
+		// to our consumers.
+		if ((renderer !== null) && (index !== null)) {
 			return;
 		}
-		const renderer = this._getTerminalRendererById(id);
-		const terminal = new ExtHostTerminal(this._proxy, name, id, renderer ? RENDERER_NO_PROCESS_ID : undefined);
-		this._terminals.push(terminal);
-		this._onDidOpenTerminal.fire(terminal);
+
+		// The extension host did not know about this terminal, so create extension host
+		// objects to represent them.
+		if (!index) {
+			const terminal = new ExtHostTerminal(this._proxy, name, id, renderer ? RENDERER_NO_PROCESS_ID : undefined);
+			index = this._terminals.push(terminal);
```

Comment:
index is not used after this point, so setting it here and removing the const above seems unnecessary.

Reply:
This change can be reverted. Thanks for spotting this.